File: results/1.0/PNG-Image/png-image-gemini-1.5-flash.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

typedef struct {
    uint32_t length;
    char type[4];
    uint8_t* data;
    uint32_t crc;
} Chunk;

HammerParser ChunkParser() {
    return h_seq(
        h_uint32_be(),
        h_string(4),
        h_bytes(h_uint32_be()),
        h_uint32_be(),
        h_map4(&(Chunk){0}, (HammerMapFunc) memcpy)
    );
}

typedef struct {
    uint32_t width;
    uint32_t height;
    uint8_t bit_depth;
    uint8_t color_type;
    uint8_t compression_method;
    uint8_t filter_method;
    uint8_t interlace_method;
} IHDR;

HammerParser IHDRParser() {
    return h_map(h_seq(
                     h_uint32_be(),
                     h_uint32_be(),
                     h_uint8(),
                     h_uint8(),
                     h_uint8(),
                     h_uint8(),
                     h_uint8(),
                     h_uint8()), (HammerMapFunc) memcpy);
}

HammerParser PNGParser() {
    return h_seq(
        h_string("89PNG\r\n\x1a\n"),
        h_map(ChunkParser(), (HammerMapFunc) memcpy),
        h_many(h_map(ChunkParser(), (HammerMapFunc) memcpy)),
        h_map(ChunkParser(), (HammerMapFunc) memcpy),
        &(PNG){0}
    );
}

typedef struct {
    char signature[8];
    Chunk ihdr;
    Chunk* idat;
    size_t idat_count;
    Chunk iend;
} PNG;

int main(int argc, char** argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <png_file>\n", argv[0]);
        return 1;
    }

    FILE* fp = fopen(argv[1], "rb");
    if (fp == NULL) {
        perror("Error opening file");
        return 1;
    }

    fseek(fp, 0, SEEK_END);
    long fsize = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    uint8_t* buffer = (uint8_t*)malloc(fsize);
    fread(buffer, 1, fsize, fp);
    fclose(fp);

    HammerParser parser = PNGParser();
    HammerResult result = hammer_parse(parser, buffer, fsize);

    if (result.success) {
        PNG* png = (PNG*)result.value;
        printf("PNG parsed successfully!\n");
        IHDR* ihdr_data = (IHDR*)png->ihdr.data;
        printf("Width: %u\n", ihdr_data->width);
        printf("Height: %u\n", ihdr_data->height);

        free(png->ihdr.data);
        for (size_t i = 0; i < png->idat_count; i++) {
            free(png->idat[i].data);
        }
        free(png->idat);
        free(png->iend.data);
        free(png);
    } else {
        fprintf(stderr, "Parsing failed at offset %zu: %s\n", result.offset, result.error);
        free(result.value);
    }
    free(buffer);
    return 0;
}

File: results/1.0/PNG-Image/png-image-gemini-1.5-flash.rs
use std::env;
use std::fs;
use std::error::Error;
use nom::{
    IResult,
    bytes::complete::{take, tag},
    number::complete::{u32, u16},
    combinator::{map, map_res, all_consuming},
    sequence::tuple,
    multi::count,
};

#[derive(Debug)]
struct PngHeader {
    signature: [u8; 8],
}

#[derive(Debug)]
struct IHDR {
    width: u32,
    height: u32,
    bit_depth: u8,
    color_type: u8,
    compression_method: u8,
    filter_method: u8,
    interlace_method: u8,
}

#[derive(Debug)]
struct Chunk {
    length: u32,
    type_: [u8; 4],
    data: Vec<u8>,
    crc: u32,
}


fn png_signature(input: &[u8]) -> IResult<&[u8], PngHeader> {
    map(tag(b"\x89PNG\r\n\x1a\n"), |_| PngHeader { signature: *b"\x89PNG\r\n\x1a\n" })(input)
}

fn ihdr(input: &[u8]) -> IResult<&[u8], IHDR> {
    map(tuple((u32, u32, u8, u8, u8, u8, u8)), |(width, height, bit_depth, color_type, compression_method, filter_method, interlace_method)| IHDR {
        width,
        height,
        bit_depth,
        color_type,
        compression_method,
        filter_method,
        interlace_method,
    })(input)
}

fn chunk(input: &[u8]) -> IResult<&[u8], Chunk> {
    map(tuple((u32, take(4usize), |i| take(i)(i), u32)), |(length, type_, data, crc)| Chunk {
        length,
        type_: type_.try_into().unwrap(),
        data: data.to_vec(),
        crc,
    })(input)
}

fn png(input: &[u8]) -> IResult<&[u8], (PngHeader, Vec<Chunk>)> {
    let (input, header) = png_signature(input)?;
    let (input, chunks) = many0(chunk)(input)?;
    Ok((input, (header, chunks)))
}

fn many0<I, O, E: nom::error::ParseError<I>>(f: impl Fn(I) -> IResult<I, O, E>) -> impl Fn(I) -> IResult<I, Vec<O>, E> {
    nom::multi::many0(f)
}


fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: png_parser <filename>");
        return Ok(());
    }

    let filename = &args[1];
    let file_contents = fs::read(filename)?;


    match all_consuming(png)(&file_contents) {
        Ok((_, (header, chunks))) => {
            println!("PNG Header: {:?}", header);
            for chunk in chunks {
                println!("Chunk: {:?}", chunk);
            }
        }
        Err(e) => {
            println!("Error parsing PNG: {}", e);
        }
    }

    Ok(())
}
